controllers/replace_failed_process_groups_test.go

```diff
@@ -63,6 +69,628 @@ var _ = Describe("replace_failed_process_groups", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			result, err := reconcileClusterWithCustomRequeueLimit(cluster, 1)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(result.Requeue).To(BeTrue())
+			Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{}))
+
+			Eventually(func() *corev1.Pod {
+				result, err = reconcileCluster(cluster)
+				Expect(err).NotTo(HaveOccurred())
+				// target pod should have been removed by reconciliation
+				return getPodByProcessGroupID(cluster, internal.GetProcessClassFromMeta(cluster, podOnTaintedNode.ObjectMeta), internal.GetProcessGroupIDFromMeta(cluster, podOnTaintedNode.ObjectMeta))
+			}).WithTimeout(time.Duration(cluster.GetTaintReplacementTimeSeconds()*3) * time.Second).WithPolling(1 * time.Second).Should(BeNil())
+
+			Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{}))
+		})
+
+		It("should remove all pods on tainted nodes when many nodes are tainted", func() {
+			concurrentTaints := 2
+			Expect(len(allPods)).To(BeNumerically(">", concurrentTaints))
+			taintedNodesIndex := map[int]struct{}{}
+			taintedNodes := []*corev1.Node{}
+			for len(taintedNodesIndex) < concurrentTaints {
+				taintedNodesIndex[rand.Intn(len(allPods))] = struct{}{}
+			}
+			for key := range taintedNodesIndex {
+				curPod := allPods[key]
+				curNode := &corev1.Node{
+					ObjectMeta: metav1.ObjectMeta{Name: curPod.Spec.NodeName},
+				}
+				taintedNodes = append(taintedNodes, curNode)
+				log.Info("Choose pod to taint", "Pod", curPod.Name, "Node", curNode.Name)
+			}
+			for i, taintedNode := range taintedNodes {
+				taintedNode.Spec.Taints = []corev1.Taint{
+					{
+						Key:       taintKeyMaintenance,
+						Value:     "rack_maintenance",
+						Effect:    corev1.TaintEffectNoExecute,
+						TimeAdded: &metav1.Time{Time: time.Now().Add(-1 * time.Second * time.Duration(taintKeyMaintenanceDuration+1))},
+					},
+				}
+				err = k8sClient.Update(ctx.TODO(), taintedNode)
+				Expect(err).NotTo(HaveOccurred())
+				log.Info("Taint node", "Index", i, "Node name", taintedNode.Name, "Node taints", taintedNode.Spec.Taints)
+			}
+			// Replace all tainted nodes in one reconciliation loop
+			cluster.Spec.AutomationOptions.MaxConcurrentReplacements = &concurrentTaints
+			err = k8sClient.Update(ctx.TODO(), cluster)
+			Expect(err).NotTo(HaveOccurred())
+
+			time.Sleep(time.Second * time.Duration(*cluster.Spec.AutomationOptions.Replacements.TaintReplacementTimeSeconds+1))
+			retry := len(taintedNodes) * 2 // Hack: Ensure reconciler replaces all tainted nodes
+			for {                          // re-run reconcileCluster up to retry times, assuming each reconciliation replaces at least one pod
+				result, err := reconcileCluster(cluster)
+				Expect(err).NotTo(HaveOccurred())
+				if !result.Requeue || retry <= 0 {
+					break
+				}
+				time.Sleep(time.Microsecond * time.Duration(500)) // Removing this will cause test failure because not all tainted pods are removed
```

Comment:
Agree. I haven't figured out why with confidence. I `think` it's because it cannot reconcile quickly enough because cluster has to remove multiple pods which hits the max_concurrent_replacement limit and slows things down.